app/pages/OrganizationEditPage.tsx

```diff
@@ -118,7 +126,9 @@ const applyChanges = (
   return newTransformedItems;
 };
 
-function getDiffObject(curr, orig) {
+/** Return an object that contains only the key-value pairs in `curr` that are different than `orig`.
+ */
+function getDiffObject<T extends {}>(curr: T, orig: T): Partial<T> {
```

Comment:
I have some idea of how this "T" pattern works and I think I have even used it but my memory is a little fuzzy about details. Do you have a resource you can link offhand that explains it? I am Googling a bit but perhaps there is something out there that you can vouch for.

Reply:
Yeah, the most common name for this feature across most languages is [generics](https://en.wikipedia.org/wiki/Generic_programming), and it's the official name in TypeScript as well. (The more academic name for it is [parametric polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism), in case that term ever pops up.) Here are [TypeScript's docs on Generics](https://www.typescriptlang.org/docs/handbook/2/generics.html). One thing to keep in mind about TypeScript's implementation that is different than in most languages is that the TypeScript implementation does not affect the actual behavior of the code, since a "TypeScript compiler" basically just strips away all the TypeScript syntax. In most statically typed languages with generics, the concrete value of the generic type can influence which method you actually call.

`T` in this example would be called a _type parameter_, and it is similar to normal function parameters except that it only holds types and it generally can only be used in places where types are allowed, like after a `:` in a variable/parameter declaration, in a `as ...` type assertion, or when explicitly passing a type parameter to a function. Normally you don't have to explicitly pass in type parameters because they get inferred from the types of the arguments.

The `extends ...` is a type constraint, which says that `T` must be compatible with the type on the right-hand side of `extends`. Without writing the `extends`, you would normally read it as "for any `T`", and with the `extends` you would read it as "for any `T` that extends ...".

The difference between `getDiffObject<T extends {}>(curr: T, orig: T)` and `getDiffObject(curr: {}, orig: {})` is that in the latter, `curr` and `orig` can be any object type (note that things like `number` and `string` are not considered objects in JavaScript), but we aren't guaranteeing that they are necessarily the same object type. With the type parameter, we are saying that `curr` and `orig` must be the same type `T`, and `T` can be any type that extends `{}` (a.k.a. any object type).

One last thing to mention is that there are unfortunately conflicts in TypeScript's syntax and React's JSX syntax, since the `<>` can sometimes be mistaken for HTML tags. This specifically affects arrow functions, since the `<>` would go immediately after the `=` and before the opening parentheses of the parameter list, e.g. `const foo = <T>(x: T) => ...`. To be able to use type parameters on arrow functions, you have to add a bit of superfluous code to allow the TypeScript compiler disambiguate the two. I've been using `const foo = <T extends unknown>(x: T)` as the workaround, since that's what I originally learned, but it seems like the [modern recommendation is to just put a comma after the type parameter](https://stackoverflow.com/a/45576880), e.g. `const foo = <T,>(x: T)`.

Reply:
Awesome, thank you for the thorough explanation! That is also good to know about the syntax clash between TS and React